cmd/cert-checker/main.go

```diff
@@ -0,0 +1,217 @@
+	var certs []core.Certificate
+	_, err := c.dbMap.Select(
+		&certs,
+		"SELECT * FROM certificates WHERE issued > :issued",
+		map[string]interface{}{"issued": time.Now().Add(-time.Hour * 24 * 90)},
+	)
+	if err != nil {
+		return err
+	}
+	c.certs = make(chan core.Certificate, len(certs))
+	for _, cert := range certs {
+		c.certs <- cert
+	}
+	// Close channel so range operations won't block when the channel empties out
+	close(c.certs)
+	return nil
+}
+
+func (c *certChecker) processCerts(wg *sync.WaitGroup) {
+	for cert := range c.certs {
+		// DEBUG
+		fmt.Println("CERT:", cert.Serial)
+
+		problems := c.checkCert(cert)
+		valid := len(problems) == 0
+		c.issuedReport.Entries[cert.Serial] = reportEntry{Valid: valid, Problems: problems}
+		if !valid {
+			atomic.AddInt64(&c.issuedReport.BadCerts, 1)
+		} else {
+			atomic.AddInt64(&c.issuedReport.GoodCerts, 1)
+		}
+	}
+	wg.Done()
+}
+
+func (c *certChecker) checkCert(cert core.Certificate) (problems []string) {
+	// Check digests match
+	if cert.Digest != core.Fingerprint256(cert.DER) {
+		problems = append(problems, "Stored digest doesn't match certificate digest")
+	}
+
+	// Parse certificate
+	parsedCert, err := x509.ParseCertificate(cert.DER)
+	if err != nil {
+		problems = append(problems, fmt.Sprintf("Couldn't parse stored certificate: %s", err))
+	} else {
+		// Check we have the right expiration time
+		if parsedCert.NotAfter != cert.Expires {
+			problems = append(problems, "Stored expiration doesn't match certificate NotAfter")
+		}
+		// Check basic constraints are set
+		if !parsedCert.BasicConstraintsValid {
+			problems = append(problems, "Certificate doesn't have basic constraints set")
+		}
+		// Check the cert isn't able to sign other certificates
+		if parsedCert.IsCA {
+			problems = append(problems, "Certificate can sign other certificates")
+		}
+		// Check the cert has the correct validity period
+		if parsedCert.NotAfter.Sub(cert.Issued) > (time.Hour * 24 * 90) {
+			problems = append(problems, "Certificate has a validity period longer than 90 days")
+		}
+		// Check that the PA is still willing to issue for each name in DNSNames + CommonName
+		for _, name := range append(parsedCert.DNSNames, parsedCert.Subject.CommonName) {
+			if err = c.pa.WillingToIssue(core.AcmeIdentifier{Type: core.IdentifierDNS, Value: name}); err != nil {
+				problems = append(problems, fmt.Sprintf("Policy Authority isn't willing to issue for %s: %s", name, err))
+			}
+		}
+		// Check the cert has the correct key usage extensions
+		if !core.CmpExtKeyUsageSlice(parsedCert.ExtKeyUsage, []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth}) {
+			problems = append(problems, "Certificate has incorrect key usage extensions")
+		}
+	}
+	return problems
+}
+
+func main() {
+	app := cmd.NewAppShell("cert-checker")
+	app.App.Flags = append(app.App.Flags, cli.IntFlag{
+		Name:  "workers",
+		Value: 5,
+		Usage: "The number of cocurrent workers used to process certificates",
```

Comment:
*concurrent.

Should we use the value of GOMAXPROCS for this? I think Golang tries do do some smart detection of number of CPUs.